src/CatalogApplication/Controllers/TagController.cs

```diff
@@ -1,130 +1,174 @@
+        /// <param name="userId">The userId to look for the tag</param>
+        /// <returns>A boolean indicating whether a tag is created or not</returns>
+        [HttpGet("{noteId}/users/{userId}")]
+        public async Task<List<Tag>> GetTagsByNoteAndUser(string noteId, string userId)
+        {
+            return await GetTags($"SELECT * FROM c WHERE c.noteId={noteId} AND c.userId={userId}");
+        }
+
+        /// <summary>
+        /// Create a new tag. Pass everything other than id.
+        /// </summary>
+        /// <returns>OK if successful otherwise bad request</returns>
+        [HttpPost]
+        public async Task<IActionResult> CreateNewTag(Tag tag)
+        {
+            try
+            {
+                if (ModelState.IsValid)
+                {
+                    tag.id = System.Guid.NewGuid().ToString();
+                    await _dbService.tagContainer.CreateItemAsync(tag);
+                }
+                return Ok();
+            }
+            catch
+            {
+                return BadRequest();
+            }
+        }
+
+        /// <summary>
+        /// Updates a tag based off of id. 
+        /// </summary>
+        /// <param name="id">The id to look for the specific tag</param>
+        /// <param name="tag">the updated version of the tag</param>
+        /// <returns></returns>
+        [HttpPut("{id}")]
+        [ActionName("Edit")]
+        public async Task<ActionResult> UpdateTag(string id, Tag tag)
+        {
+            if (id == null)
+            {
+                return BadRequest();
+            }
+            if (ModelState.IsValid)
+            {
+                try
+                {
+                    await _dbService.tagContainer.UpsertItemAsync<Tag>(tag, new PartitionKey(id));
+                    return Ok();
+                }
+                catch
+                {
+                    return BadRequest();
+                }
+
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Deletes all the tags associated with a specific noteId. Useful for when a note
+        /// is deleted.
+        /// </summary>
+        /// <param name="noteId">The Note ID associated with the tag to be deleted</param>
+        /// <returns>Returns a list of tags that have been deleted</returns>
+        [HttpDelete("{noteId}")]
+        public async Task<List<Tag>> DeleteTags(string noteId)
+        {
+            QueryDefinition queryDef = new QueryDefinition($"SELECT * FROM c WHERE c.noteId={noteId}");
+            FeedIterator<Tag> iterator = _dbService.tagContainer.GetItemQueryIterator<Tag>(queryDef);
+
+            List<Tag> deletedTags = new List<Tag>();
+
+            while (iterator.HasMoreResults)
+            {
+                var resultSet = await iterator.ReadNextAsync();
+                foreach (Tag tag in resultSet)
+                {
+                    await DeleteTag(tag.id);
```

Comment:
If this operation fails you don't want to go ahead and say that the tag was deleted.